src/livestatus.c

```diff
@@ -0,0 +1,463 @@
+      .neb_callbacks_rate = strtod(fields[14], NULL),
+      .requests = strtoull(fields[15], NULL, 10),
+      .requests_rate = strtod(fields[16], NULL),
+      .service_checks = strtoull(fields[17], NULL, 10),
+      .service_checks_rate = strtod(fields[18], NULL),
+  };
+
+  return status;
+} /* livestatus_status_t c_to_livestatus_status */
+
+static int ls_init(void) {
+  livestatus_obj.socket_file = "/var/cache/naemon/live";
+  livestatus_obj.max_retry = 20;
+  livestatus_obj.backoff_sec = 1;
+
+  return 0;
+} /* int ls_init */
+
+static int ls_config(const char *key, const char *value) /* {{{ */
+{
+  if (strcasecmp("LivestatusSocketFile", key) == 0) {
+    char *socket_file = strdup(value);
+    if (socket_file == NULL) {
+      ERROR("livestatus plugin: strdup failed: %s", STRERRNO);
+      return 1;
+    }
+
+    livestatus_obj.socket_file = socket_file;
+  }
+
+  if (strcasecmp("OnFailureMaxRetry", key) == 0) {
+    char *max_retry_str = strdup(value);
+    if (max_retry_str == NULL) {
+      ERROR("livestatus plugin: strdup failed: %s", STRERRNO);
+      return 1;
+    }
+
+    livestatus_obj.max_retry = atoi(value);
+  }
+
+  if (strcasecmp("OnFailureBackOffSeconds", key) == 0) {
+    char *backoff_str = strdup(value);
+    if (backoff_str == NULL) {
+      ERROR("livestatus plugin: strdup failed: %s", STRERRNO);
+      return 1;
+    }
+
+    livestatus_obj.backoff_sec = atoi(value);
+  }
+
+  return 0;
+} /* static int ls_config */
+
+static int unix_connect(const char *sockfile, int *sockfd) {
+  struct sockaddr_un sun;
+  int rc = -1;
+  int sfd = -1;
+
+  sfd = socket(AF_UNIX, SOCK_STREAM, 0);
+  if (sfd < 0) {
+    ERROR("livestatus plugin: creating client socket: %s", STRERRNO);
+    return -1;
+  }
+
+  memset(&sun, 0x0, sizeof(sun));
+  sun.sun_family = AF_UNIX;
+  strcpy(sun.sun_path, sockfile);
+
+again:
+  rc = connect(sfd, (struct sockaddr *)&sun, sizeof(sun));
+  if (rc < 0) {
+    if (errno != EINTR) {
+      ERROR("livestatus plugin: connect to socket file %s: %s", sockfile,
+            STRERRNO);
+      return -1;
+    }
+    goto again;
+  }
+
+  *sockfd = sfd;
+
+  return 0;
+} /* unix_connect */
+
+static int ls_parse(const char *lresponse, livestatus_status_t *lstatus) {
+  int rc = -1;
+  int i = 0;
+  char **fields = NULL;
+  char *pchar = NULL;
+
+  fields = malloc(sizeof(char **) * LIVESTATUS_EXPECTED_FIELDS_RESP_NB);
+  if (fields == NULL) {
+    ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+    return -1;
+  }
+
+  for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+    fields[j] = malloc(sizeof(char *) * 32);
+    if (fields[j] == NULL) {
+      ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+      rc = -1;
+      goto free_all_fields;
+    }
+
+    memset(fields[j], 0x0, 32);
+  }
+
+  pchar = (char *)lresponse;
+  char *wstart = (char *)lresponse;
+  int exit = 0;
+  while (!exit) {
+    if (*pchar == '\0' || *pchar == '\n') {
+      exit = 1;
+    } else {
```

Comment:
'else if (*pchar != ';')' to avoid one unnecessary level of nesting

Reply:
Fixed.